### DISCUSSION TITLE
A GitHub action for setting up and running Auto

### DISCUSSION THREAD

**Discussion by laughedelic** [intuit/auto#1958] [Ideas]
Hi! I want to use Auto on GitHub actions in non-npm projects. I need to install Auto standalone binary for that and while this is easy to do with a little shell-script, it would be better to have a reusable GitHub action. Here are existing actions I found on the marketplace (searching anything related to Auto is really hard, btw 😞):

- https://github.com/sugarmanz/auto-native-setup-action: installs Auto binary from GitHub releases (Linux-only)
- https://github.com/terradatum/auto-action: doesn't install Auto (?), only wraps it's npx-based CLI

The first one works and does basically what I want, but it's not portable (which is ok for my current use-case, but generally is an arbitrary limitation) and seems to be not properly released (there are a few tags marked as pre-release).

---

## Proposal

I want to write an action that will install Auto (linux/macos/windows), optionally run it with given arguments and expose output as the action output, so it's easier to integrate with other actions.

Here is some example usages:

### Simple setup

**Update:** This is now implemented: https://github.com/laughedelic/setup-auto

```yaml
- name: Setup Auto
  uses: intuit/setup-auto@v1
```

Would be equivalent to running steps like this:

```shell
- name: Set up GitHub credentials for Auto
  run: |
    echo "GH_TOKEN=${{ github.token }}" >> $GITHUB_ENV
    git config user.name "github-actions[bot]"
    git config user.email "41898282+github-actions[bot]@users.noreply.github.com"

- name: Install Auto binary
  run: |
    platform=$(...)
    gh release download -R intuit/auto -p auto-$platform.gz
    gunzip -c ./auto-$platform.gz > /usr/bin/auto
    chmod a+x /usr/bin/auto

- name: Check Auto installation
  run: auto --help
```

Natually, git name/email/token can be optionally configured.

### Running a command

```yaml
- name: Setup Auto
  uses: intuit/setup-auto@v1
  with:
    command: shipit
```

or with any other CLI options:

```yaml
...
  with:
    command: shipit --dry-run --no-changelog
```

The motivation of running Auto from the action instead of having a separate step like `run: auto shipit` is that from the action it's easier to expose information from the release as outputs. Then next steps (e.g. deployment) can use that information (e.g. released version). 

From this angle it might be better to implement action in TypeScript and use Auto's API directly. Then maybe we don't need to use the platform-dependent binary.

---

What do you think about this?

**Comment by laughedelic** [comment]
I created a simple version that just installs (pre-packaged) Auto and configures Git author/credentials: https://github.com/laughedelic/setup-auto

Would you consider moving it to the @intuit org? I can maintain it, but don't want the "ownership" of this action, just want an easy official way to use Auto on GitHub Actions.

**Reply by hipstersmoothie** [reply]
ooo I really like how this turned out. Unfortunately I'm not at Intuit anymore so I don't think I can add repos or member really. I think it would be cool to include a link to the action somewhere in our docs though

**Reply by laughedelic** [reply]
Sure, I will update this page: https://intuit.github.io/auto/docs/build-platforms/github-actions

Maybe @adierkens or someone else from Intuit can help with transferring the repo, if it's a desired outcome?

**Reply by just-be-dev** [reply]
We do have the @auto-it organization we could add it to that's more open source. Just a bunch of test stuff there right now.

**Reply by laughedelic** [reply]
@zephraph that sounds great! having auto in the name is even better 😄 

I just added you to the repo, so that you will be able to do the transfer. I can't do it because I'm not a member of @auto-it.

**Reply by just-be-dev** [reply]
Looks like I don't have permissions to move, but I did invite you to the @auto-it org.

**Reply by laughedelic** [reply]
it's done! https://github.com/auto-it/setup-auto
thanks @zephraph.
